### PR TITLE
Match sites to NHDPlusv2 flowlines

### LINKED ISSUE #3: Match temperature sites to NHDv2 flowlines
We'll need a target that matches the DRB temperature site locations to NHDv2 flowline reaches.

### PR DESCRIPTION
**Pull request by lekoenig**
This PR adds new targets representing the temperature site locations (`p1_drb_temp_sites_shp`, `p1_drb_temp_sites_sf`) and data  (`p1_drb_temp_obs_csv`, `p1_drb_temp_obs`), and then matches the temperature sites to NHDPlusv2 flowlines (`p2_drb_temp_sites_w_segs`). 

The function `match_sites_to_reaches` includes an argument, `use_NHM_match`, that dictates whether we should restrict the search for NHDv2 matches based on the previous match that was done as part of the temperature forecasting project. In that project, sites were matched to minimize the "fish distance" between a site and the downstream end of an NHM segment (since the PRMS temperature outputs used for pre-training most closely represent conditions at the downstream end of each segment). As a result, the matched NHM segment is not always the segment that is closest to a given site. 

As the default, `use_NHM_match` is currently set to `TRUE`, so this means that we'll match the COMIDs based on the subset of NHDv2 reaches that overlap the previously-matched NHM segment. Here's an example of what that looks like for a subset of sites previously matched to segidnat 1435:

sites2

On the one hand, restricting our NHDv2 matches this way provides a closer match to how the models using the NHM fabric are trained. However, the above example illustrates that this procedure results in many sites that are matched to a COMID as far as 10 km upstream (in the most extreme case) as opposed to COMID's that are directly adjacent to those sites. I'd appreciate any thoughts you have on whether this approach (`use_NHM_match = TRUE`) aligns with our downscaling and/or groundwater-focused objectives, or whether it makes more sense for us to match the sites to NHDv2 reaches based on a simple "bird distance" cutoff (`use_NHM_match = FALSE`). 

Closes #3

### CODE REVIEW

**Comment by lekoenig** [comment]
When thinking about how to best match sites to reaches, we'd like to be able to compare model runs using the NHDv2 fabric with those using the NHM fabric. As mentioned above, this is slightly complicated by the fact that temperature modeling work in the DRB thus far has used a specific algorithm to match sites to reaches: reaches for which the downstream endpoint (vertex) is close to the site location are preferred over the nearest match using simple 'bird' distances. One thing @janetrbarclay and I discussed was whether we could emulate that procedure but using the NHDPlusv2 flowlines instead.

**Comment by lekoenig** [comment]
@janetrbarclay, in https://github.com/USGS-R/drb-gw-hw-model-prep/pull/7/commits/b41b11ab4cc1bed6de7b2a6ac64ebb6414ae162b I implemented the code changes we discussed earlier, i.e. emulate the process used by prior temperature work when matching sites to nhdv2 flowlines, preferring reaches for which the downstream vertex (endpoint) is close to the site location. 

Here's what the output in `p2_drb_temp_sites_w_segs` now looks like for the example reaches above, where the sites are color-coded based on the COMID they get matched to in `subset_closest_nhd()`. So a couple of the teal sites are closest to comid `1748541` but get matched to `1748709` because they are closer to the downstream vertex of that reach than they are to the downstream vertex of `1748541`. 
site_matches

Let me know what you think about this approach. I've left the original implementation in this PR for now (renamed as `p2_drb_temp_sites_w_segs_use_NHM_match`), but will probably opt to delete one of the site-to-segment targets before merging.

**Comment by janetrbarclay** [comment]
@lekoenig I think that approach sounds good. It is consistent with the prior philosophy of matching to the nearest segment endpoint and it aggregates the observations at the NHD resolution rather than the NHM. I'll take a look at the code this morning and complete my review.

File: 2_process/src/subset_closest_nhd.R [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  # Find the nearest NHDv2 segment to each point and return the 
+  # 'bird distance' between the site and the nearest segment
+  # [Alison] for guidance used for this implementation, see 
+  # https://gis.stackexchange.com/questions/288570/find-nearest-point-along-polyline-using-sf-package-in-r
+  nearest_segment <- nhd_lines[sf::st_nearest_feature(sites, nhd_lines), ]
+  bird_dist_to_seg_m <- sf::st_length(
+    sf::st_nearest_points(sites, nearest_segment, pairwise = TRUE))
+  
+  # Find the nearest NHDv2 vertex to each point and return the
+  # 'fish distance' between the site and the nearest vertex
```
[the remark this answers is not in the document]
Comment (lekoenig) [reply 2_process/src/subset_closest_nhd.R:96]: Good catch! I'll make that change.

**Pull request merged**